modules/calib3d/src/calibinit.cpp

```diff
@@ -300,213 +464,258 @@ int cvFindChessboardCorners( const void* arr, CvSize pattern_size,
         {
-            return 0;
+            IplImage _img;
+            cvGetImage(img, &_img);
+            check_chessboard_result = cvCheckChessboard(&_img, pattern_size);
+            if(check_chessboard_result <= 0)
+            {
+                return 0;
+            }
         }
     }
 
     // Try our standard "1" dilation, but if the pattern is not found, iterate the whole procedure with higher dilations.
     // This is necessary because some squares simply do not separate properly with a single dilation.  However,
     // we want to use the minimum number of dilations possible since dilations cause the squares to become smaller,
     // making it difficult to detect smaller squares.
-    for( k = 0; k < 6; k++ )
+    for( dilations = min_dilations; dilations <= max_dilations; dilations++ )
     {
-        int max_quad_buf_size = 0;
-        for( dilations = min_dilations; dilations <= max_dilations; dilations++ )
-        {
-            if (found)
-                break;      // already found it
+      if (found)
+        break;      // already found it
 
-            cvFree(&quads);
-            cvFree(&corners);
+      cvFree(&quads);
+      cvFree(&corners);
 
-            /*if( k == 1 )
-            {
-                //Pattern was not found using binarization
-                // Run multi-level quads extraction
-                // In case one-level binarization did not give enough number of quads
-                CV_CALL( quad_count = icvGenerateQuadsEx( &quads, &corners, storage, img, thresh_img, dilations, flags ));
-                PRINTF("EX quad count: %d/%d\n", quad_count, expected_corners_num);
-            }
-            else*/
-            {
-                // convert the input grayscale image to binary (black-n-white)
-                if( flags & CV_CALIB_CB_ADAPTIVE_THRESH )
-                {
-                    int block_size = cvRound(prev_sqr_size == 0 ?
-                        MIN(img->cols,img->rows)*(k%2 == 0 ? 0.2 : 0.1): prev_sqr_size*2)|1;
-
-                    // convert to binary
-                    cvAdaptiveThreshold( img, thresh_img, 255,
-                        CV_ADAPTIVE_THRESH_MEAN_C, CV_THRESH_BINARY, block_size, (k/2)*5 );
-                    if (dilations > 0)
-                        cvDilate( thresh_img, thresh_img, 0, dilations-1 );
-                }
-                else
-                {
-                    // Make dilation before the thresholding.
-                    // It splits chessboard corners
-                    //cvDilate( img, thresh_img, 0, 1 );
+      int max_quad_buf_size = 0;
 
-                    // empiric threshold level
-                    double mean = cvAvg( img ).val[0];
-                    int thresh_level = cvRound( mean - 10 );
-                    thresh_level = MAX( thresh_level, 10 );
+      //USE BINARY IMAGE COMPUTED USING icvBinarizationHistogramBased METHOD
+      cvDilate( thresh_img_new, thresh_img_new, 0, 1 );
```

Comment:
cvFindContours() (from icvGenerateQuads) modifies image, which becomes changed for next dilation iterations.

#11589